syncer/syncer.go

```diff
@@ -0,0 +1,218 @@
+	lookupTable map[crypto.Hash]struct{}
+)
+
+// New returns a new Syncer with the given parameters.
+func New(ctx context.Context, blocker *blocker.Blocker, skydAPI skyd.API, db *database.DB, portalURLs []string, logger *logrus.Logger) (*Syncer, error) {
+	if ctx == nil {
+		return nil, errors.New("no context provided")
+	}
+	if blocker == nil {
+		return nil, errors.New("no blocker provided")
+	}
+	if db == nil {
+		return nil, errors.New("no DB provided")
+	}
+	if logger == nil {
+		return nil, errors.New("no logger provided")
+	}
+	if skydAPI == nil {
+		return nil, errors.New("no Skyd API provided")
+	}
+	s := &Syncer{
+		staticBlocker:    blocker,
+		staticCtx:        ctx,
+		staticDB:         db,
+		staticLogger:     logger,
+		staticPortalURLs: portalURLs,
+		staticSkydAPI:    skydAPI,
+	}
+	return s, nil
+}
+
+// Start launches a background task that periodically syncs the blocklists of
+// the preconfigured portals with the blocklist of the local skyd instance.
+func (s *Syncer) Start() {
+	// convenience variables
+	logger := s.staticLogger
+
+	// escape early if the syncer has no portal urls configured
+	if len(s.staticPortalURLs) == 0 {
+		logger.Infof("syncer is not being started because no portal URLs have been defined")
+		return
+	}
+
+	// start the sync loop
+	go s.threadedSyncLoop()
+}
+
+// threadedSyncLoop holds the main sync loop
+func (s *Syncer) threadedSyncLoop() {
+	// convenience variables
+	logger := s.staticLogger
+
+	for {
+		err := s.syncPortalsWithSkyd()
+		if err != nil {
+			logger.Errorf("failed to sync portals with skyd, error %v", err)
+		}
+
+		select {
+		case <-s.staticCtx.Done():
+			return
+		case <-time.After(syncInterval):
+		}
+	}
+}
+
+// blocklistFromPortal returns the blocklist for the portal at the given URL.
+func (s *Syncer) blocklistFromPortal(portalURL string) ([]crypto.Hash, error) {
+	// convenience variables
+	logger := s.staticLogger
+
+	// build the request to fetch the blocklist from the portal
+	url := fmt.Sprintf("%s/skynet/blocklist", portalURL)
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to build blocklist request")
+	}
+
+	// set headers and execute the request
+	req.Header.Set("User-Agent", "Sia-Agent")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to fetch blocklist")
+	}
+	defer func() {
+		err = resp.Body.Close()
+		if err != nil {
+			logger.Errorf("failed to close the request body, err: %v", err)
+		}
+	}()
+
+	// decode the response
+	br := struct{ Blocklist []crypto.Hash }{}
+	err = json.NewDecoder(resp.Body).Decode(&br)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to decode skyd API response")
+	}
+
+	return br.Blocklist, nil
+}
+
+// syncPortalsWithSkyd will sync the blocklist of all portals defined on the
+// syncer with the local skyd.
+func (s *Syncer) syncPortalsWithSkyd() error {
```

Comment:
This method downloads blocklists from other portals and then blocks them on the local server but it doesn't add them to the portal's blocklist itself. 

If we only block the hashes in skyd but don't put them in the DB we have a few suboptimal outcomes:
1. Each new server that you add to your cluster must rely on syncing those blocklist entries from the outside server as well, instead of fetching them from the local DB.
2. You can't get a complete list of what's blocked on your portal. Knowing what's blocked might be useful if you need to argue with law enforcement or unblock something due to a valid challenge of the original block.
3. If you're running your own malware-scanning service, you'll end up re-downloading and re-scanning something somebody already scanned.
4. If you whitelist some legitimate content that was erroneously blocked, those blocks will be dropped from your local `skyd`, only to be reinstated on the next pull from an external server which still has the illegitimate block.

All in all, I don't see a benefit in not updating our own ledger of bad skylink hashes and blindly accepting what other portals have blocked without double-checking with our own list first.

Reply:
I see your point but this essentially puts into question the entire PR. I personally don't think any of the bullet points above are an issue really, but others might feel differently. The reason I did not add the hashes to the local database is that those records would be totally different from the other ones. There is no reporter, there are no tags, there is no Skylink and so on, it's just a hash. The sync is just a way for new portal operators to bootstrap their blocklist with the blocklist of a trusted portal and do so with ease.

1. Every server should indeed have the same portals defined in the environment for that local server's skyd to contain all the hashes another server would have. I don't think this is an issue really, we manage env variables from a single source of truth anyway.
2. You can get a list of blocked hashes from skyd. We are removing away from storing skylinks everywhere anyway so pulling a list of skylinks you are blocking is impossible.
3. I don't fully understand this bullet point, but you probably wouldn't want to trust the scan results of a 3rd party anyway.
4. Currently the allow list is based on skylinks. That's being moved in the second part of the skylink to hash refactor. When that is complete I will be able to check the hashes against the whitelist and check before blocking, but this bullet is a good point.

This feature is just the result of Matt saying "it would be nice to have a button somewhere to sync the blocklist from portal A to portal B". I think there are plenty of use cases where as a portal operator you would be fine with syncing the blocklist from a trusted friend portal, and if you're not you simply don't do it.

Reply:
2. You can get a list of blocked hashed *per server*, sure, but you'd need to query each server separately and then merge the results. Anywhere where I'm talking about skylinks I mean their hash representation.
3. The idea here is that we are already blocking some skylink because it's been blocked by a portal we sync with. But since it's been requested and it's not in the DB, we will still run it through our malware-scanner, wasting resources. And the best part here is that regardless of the result of the scan, the hash will still be blocked anyway, because we don't check those hashed against our allowlist.

I've heard this feature described as "a way to bootstrap a portal" and I think that's fine. But if it's means as a bootstrapping tool then:
1. It shouldn't be run routinely, keeping in sync with the other portal.
2. It should still store the hashes in the DB because that's what bootstrapping usually means - getting a solid starting point on which to build. If we don't do that and we later add another server to our portal but the portal we synced from is down or dead, the new server will have a different blocklist from the rest of our portal. That's not ideal from the standpoint of predictable behaviour of our own servers.

Reply:
I can't update the PR in a way that would allow me to resolve your comment. Adding them to the database from the blocker is not the desired approach in my opinion because we would not have a reporter, nor a list of tags. If we would follow your reasoning we would need an entirely different tool that imports the entire mongo collection or keeps those in sync. I'll wait for comments from either @MSevey or @ChrisSchinnerl (or both) to proceed.

Reply:
I think we need to establish what we want to achieve first.
1. A one-time bootstrap - in that case this PR may be suboptimal because we don't bootstrap our own db.
2. A way to keep up-to-date with a trusted portal's list. Which will usually be used by people who don't feed their own db and just want to mirror a different portal.

Reply:
Matt asked for a one-time bootstrap button. I just figured I was extending the idea in a good direction. 

I felt syncing > bootstrapping, because the end result will be that if you configure a sync, your portal will benefit from having abusive skylinks blocked, even though they were not reported to you directly. In the end, we want to just block CSAM links on as many portals as fast as possible right.

Reply:
ok resolved on Discord - we'll add a route to the blocker (partially) exposing the blocklist collection
instead of syncing with a server's blocklist, we'll sync with a portal's blocklist, which is way better
and I fully agree with merging those db entities

Reply:
marked as draft as I need to make some changes, resolving this comment